
syntax = "proto3";

// Communication from the debugger adapter to unreal. This is effectively a mapping
// of the commands defined in the "The Callback" section of the debugger interface.
// https://docs.unrealengine.com/udk/Three/DebuggerInterface.html#The%20Callback

package unreal_debugger.commands;

message AddBreakpoint
{
    string class_name = 1;
    int32 line_number = 2;
}

message RemoveBreakpoint
{
    string class_name = 1;
    int32 line_number = 2;
}

message AddWatch
{
    string var_name = 1;
}

message RemoveWatch
{
    string var_name = 1;
}

message ClearWatch
{
}

message ChangeStack
{
    int32 stack_id = 1;
}

message SetDataWatch
{
    string var_name = 1;
}

message BreakOnNone
{
    bool break_value = 1;
}

message Break
{
}

message StopDebugging
{
}

message Go
{
}

message StepInto
{
}

message StepOver
{
}

message StepOutOf
{
}

message Command
{
    enum Kind {
        Unknown = 0;
        AddBreakpoint = 1;
        RemoveBreakpoint = 2;
        AddWatch = 3;
        RemoveWatch = 4;
        ClearWatch = 5;
        ChangeStack = 6;
        SetDataWatch = 7;
        BreakOnNone = 8;
        Break = 9;
        StopDebugging = 10;
        Go = 11;
        StepInto = 12;
        StepOver = 13;
        StepOutOf = 14;
    }

    Kind kind = 1;

    oneof payload {
        AddBreakpoint add_breakpoint = 2;
        RemoveBreakpoint remove_breakpoint = 3;
        AddWatch add_watch = 4;
        RemoveWatch remove_watch = 5;
        ClearWatch clear_watch = 6;
        ChangeStack change_stack = 7;
        SetDataWatch set_data_watch = 8;
        BreakOnNone break_on_none = 9;
        Break break_cmd = 10;
        StopDebugging stop_debugging = 11;
        Go go = 12;
        StepInto step_into = 13;
        StepOver step_over = 14;
        StepOutOf step_out_of = 15;
    }
}
